Turn off color depending on PY_COLORS / NO_COLOR

Currently (black 22.1.0) it is not possible to turn off colored output, in particular in the summary at the end. Please provide a simple way to turn color off, preferably via one of the environment variables PY_COLORS or NO_COLOR.

There are also dumb terminals which do not interpret ANSI escape sequences, instead showing them verbatim.

In my specific case, the env var is `TERM=dumb` (emacs compilation buffer on Linux), and as `compilation`, I am running first `black` and then `flake8`.

In the compilation buffer, lines like

    foo/bar.py:23:42: F821 undefined name 'txt'

will be colored by the editor to show the `foo/bar.py:23:42:` part and allow clicking on that part to jump to the proper place in the source file.

Unfortunately, directly above that line from `flake8` will be

    ^[[1mAll done! ✨ 🍰 ✨^[[0m
    ^[[34m10 files ^[[0mleft unchanged.

which is not very readable.

FWIW, the workaround appears to be to tell black that its stderr is not a TTY:

    black8 2>&1 | cat

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -4,7 +4,9 @@
 
 <!-- PR authors:
      Please include the PR number in the changelog entry, not the issue number -->
+
 - Add support for NO_COLOR environment variable to disable ANSI output (#5129)
+
 ### Highlights
 
 <!-- Include any especially major or disruptive changes here -->

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -4,7 +4,7 @@
 
 <!-- PR authors:
      Please include the PR number in the changelog entry, not the issue number -->
-
+- Add support for NO_COLOR environment variable to disable ANSI output (#5129)
 ### Highlights
 
 <!-- Include any especially major or disruptive changes here -->

--- a/src/black/output.py
+++ b/src/black/output.py
@@ -4,6 +4,7 @@ The double calls are for patching purposes in tests.
 """
 
 import json
+import os
 import re
 import tempfile
 from typing import Any
@@ -17,7 +18,7 @@ def _out(message: str | None = None, nl: bool = True, **styles: Any) -> None:
     if message is not None:
         if "bold" not in styles:
             styles["bold"] = True
-        message = style(message, **styles)
+        message = style_output(message, **styles)
     echo(message, nl=nl, err=True)
 
 
@@ -26,7 +27,7 @@ def _err(message: str | None = None, nl: bool = True, **styles: Any) -> None:
     if message is not None:
         if "fg" not in styles:
             styles["fg"] = "red"
-        message = style(message, **styles)
+        message = style_output(message, **styles)
     echo(message, nl=nl, err=True)
 
 
@@ -95,6 +96,8 @@ def diff(a: str, b: str, a_name: str, b_name: str) -> str:
 
 def color_diff(contents: str) -> str:
     """Inject the ANSI color codes to the diff."""
+    if not _color_enabled():
+        return contents
     lines = contents.split("\n")
     for i, line in enumerate(lines):
         if line.startswith("+++") or line.startswith("---"):
@@ -107,6 +110,16 @@ def color_diff(contents: str) -> str:
             line = "\033[31m" + line + "\033[0m"  # red, reset
         lines[i] = line
     return "\n".join(lines)
+
+
+def style_output(message: str, **styles: Any) -> str:
+    if not _color_enabled():
+        return message
+    return style(message, **styles)
+
+
+def _color_enabled() -> bool:
+    return "NO_COLOR" not in os.environ
 
 
 @mypyc_attr(patchable=True)

--- a/src/black/report.py
+++ b/src/black/report.py
@@ -6,9 +6,7 @@ from dataclasses import dataclass
 from enum import Enum
 from pathlib import Path
 
-from click import style
-
-from black.output import err, out
+from black.output import err, out, style_output
 
 
 class Changed(Enum):
@@ -94,14 +92,20 @@ class Report:
         if self.change_count:
             s = "s" if self.change_count > 1 else ""
             report.append(
-                style(f"{self.change_count} file{s} ", bold=True, fg="blue")
-                + style(f"{reformatted}", bold=True)
+                style_output(
+                    f"{self.change_count} file{s} ", bold=True, fg="blue"
+                )
+                + style_output(f"{reformatted}", bold=True)
             )
 
         if self.same_count:
             s = "s" if self.same_count > 1 else ""
-            report.append(style(f"{self.same_count} file{s} ", fg="blue") + unchanged)
+            report.append(
+                style_output(f"{self.same_count} file{s} ", fg="blue") + unchanged
+            )
         if self.failure_count:
             s = "s" if self.failure_count > 1 else ""
-            report.append(style(f"{self.failure_count} file{s} {failed}", fg="red"))
+            report.append(
+                style_output(f"{self.failure_count} file{s} {failed}", fg="red")
+            )
         return ", ".join(report) + "."

--- a/src/black/report.py
+++ b/src/black/report.py
@@ -92,9 +92,7 @@ class Report:
         if self.change_count:
             s = "s" if self.change_count > 1 else ""
             report.append(
-                style_output(
-                    f"{self.change_count} file{s} ", bold=True, fg="blue"
-                )
+                style_output(f"{self.change_count} file{s} ", bold=True, fg="blue")
                 + style_output(f"{reformatted}", bold=True)
             )
 

--- a/tests/test_black.py
+++ b/tests/test_black.py
@@ -240,6 +240,23 @@ class BlackTestCase(BlackBaseTestCase):
         self.assertIn("\033[31m", actual)
         self.assertIn("\033[0m", actual)
 
+    def test_piping_diff_with_color_respects_no_color(self) -> None:
+        source, _ = read_data("cases", "expression.py")
+        args = [
+            "-",
+            "--fast",
+            f"--line-length={black.DEFAULT_LINE_LENGTH}",
+            "--diff",
+            "--color",
+            f"--config={EMPTY_CONFIG}",
+        ]
+        with patch.dict(os.environ, {"NO_COLOR": "1"}):
+            result = BlackRunner().invoke(
+                black.main, args, input=BytesIO(source.encode("utf-8"))
+            )
+        actual = result.output
+        self.assertNotIn("\033[", actual)
+
     def test_pep_572_version_detection(self) -> None:
         source, _ = read_data("cases", "pep_572")
         root = black.lib2to3_parse(source)
@@ -794,6 +811,13 @@ class BlackTestCase(BlackBaseTestCase):
                 "2 files would be reformatted, 3 files would be left unchanged, 2"
                 " files would fail to reformat.",
             )
+
+    def test_report_respects_no_color(self) -> None:
+        report = Report()
+        report.done(Path("f1"), black.Changed.YES)
+        with patch.dict(os.environ, {"NO_COLOR": "1"}):
+            output = str(report)
+        self.assertEqual(output, unstyle(output))
 
     def test_lib2to3_parse(self) -> None:
         with self.assertRaises(black.InvalidInput):

